1196 - Fix import/export bugs and limitations; add import from zip

zifgu opened the pull request:
## Changes
- Moves import file validation and permissions filtering to the backend.
- Adds ability to export a Tapestry to a ZIP file, which will bundle the regular JSON export file with all **locally uploaded** media (videos, images, H5Ps) in the Tapestry. This method is suitable for moving an entire Tapestry to a different site.
- If there are WordPress post nodes in the Tapestry, a WXR file (the export format for WordPress posts) is downloaded alongside the JSON or ZIP file. Otherwise, only the export file is downloaded. 
- Adds ability to import an exported Tapestry ZIP. This **re-creates** all media referenced by the Tapestry on the destination site.
- Expands the ImportChangelog component to display import warnings and information to the user.
- To import a Tapestry with WordPress posts, the user should import the posts before importing the Tapestry that uses them, and the WordPress Post nodes in the Tapestry will automatically link to the new posts.
- Splits Export and Duplicate functionalities in Settings modal since Export has expanded.
- Adds a handler to Cypress allowing it to print debug logs. Not directly applicable to this PR, but may help with debugging in the future.

**Bug fixes:**
- Fixes bug that prevents importing because `group` property does not exist
- Fixes bug when importing Answer nodes (since node IDs change during import, the `activityId` field should be updated to the new node ID)
- Fixes bug where imported Tapestries do not have a title (described at https://github.com/tapestry-tool/tapestry-wp/pull/1174#issuecomment-1171773108)

## Change details
Export details:
- Nodes with problems in the media (e.g. an image wasn't found) are exported as-is, but a warning is logged. The same goes for media in the Tapestry settings (e.g. the background image). At the end, all warnings are displayed to the user.
- Each export creates a temporary ZIP file in the `wp-content/uploads/tapestry/export` folder (plus a file with the same name and no extension). The URL of this ZIP is passed back to the client to be downloaded. A daily job should automatically delete export files older than 24 hours.

ZIP file details:
- The exported ZIP file has a flat structure (no root directory).
- The ZIP file will always contain a `tapestry.json`, which is the regular JSON export file (with some minor differences, described below)
- Image and video files are saved in the ZIP with the file names at the end of their URLs. (e.g. `http://localhost:8888/wp-content/uploads/video.mp4` becomes `video.mp4`)
- H5Ps are saved in the ZIP as the corresponding export file under `wp-content/uploads/h5p/exports` (e.g. `interactive-video-35.h5p`)
- Images added via Link Preview or videos added via URL (i.e. anything not a local upload) are not saved.
- Thumbnail images are saved in their original size, not the variants generated by WordPress 
- Images inside Activity nodes are only saved if the answer type and `useImages` are enabled.
- Inside tapestry.json, the URLs to local uploads are replaced by their file names inside the ZIP
     - e.g. `http://localhost:8888/wp-admin/admin-ajax.php?action=h5p_embed&id=3` -> `my-quiz-3.h5p`
     - e.g. `http://localhost:8888/wp-content/uploads/thumbnail.jpeg` -> `thumbnail.jpeg`

Import details:
- Any nodes that failed to import (e.g. a referenced file didn't exist) are still added, but a warning is logged. The same goes for imported Settings. At the end, all warnings are displayed to the user.
- Each import creates temporary directories under the `wp-content/uploads/tapestry/import` folders for unzipping the ZIP file. These should be deleted immediately after the import. (If they are not, something is wrong.)
- Additional sizes are generated for thumbnail images, but not for other images.
- H5P caches are only rebuilt if the imported Tapestry includes H5P files and the H5P files were uploaded successfully. Rebuilding the H5P cache is time-consuming, so we want to avoid it if possible.

## Screenshot
New Export / Duplicate modal:

<img width="507" alt="image" src="https://user-images.githubusercontent.com/8981258/196546508-861a93b9-e008-4678-a15d-de7ea639fa99.png">
<img width="508" alt="image" src="https://user-images.githubusercontent.com/8981258/196546621-f1c1a434-1c30-44aa-b55d-9d06ed0f7707.png">

Import (with warnings):

https://user-images.githubusercontent.com/44929104/178632980-abc03476-9975-487e-b494-5d11bf05ee77.mp4

Export (with warnings):

<img width="507" alt="image" src="https://user-images.githubusercontent.com/8981258/196546778-57942d8a-ebed-46fc-a205-e48e32670eb1.png">

Video with outdated UI showing export with warnings: https://user-images.githubusercontent.com/44929104/178630936-3a55c808-8b0f-4d04-a2fd-715d8c95b2a8.mp4

## Testing
**To test importing WordPress post nodes**: Since we only have 1 site to test on, we have to emulate the WordPress posts being recreated.
1. Export the Tapestry
2. Open the downloaded XML file. Search for all `<wp:post_id>` tags and change the numbers to something different.
3. Delete the exported posts from your local site. Make sure to also delete them from the Trash, since WordPress will warn you that the posts still exist otherwise.
4. Go to Tools -> Import. Install the "WordPress" importer if you haven't already, and click "Run Importer". Follow the instructions to import the WordPress file.
5. Import the Tapestry ZIP file. Verify that the WordPress post nodes refer to the newly created posts.

## Issue Linkage
Closes #1196 

## PR Dependency
Depends on #1188
 
## Automated Testing
Added tests to cover several basic cases:
- Several kinds of invalid ZIP files
- Importing a ZIP file containing images and videos, with no warnings
- Importing a ZIP file containing images and videos, with some warnings
- Exporting a ZIP file
- Exporting a tapestry with WordPress posts
- Exporting a ZIP file with warnings
- Importing a ZIP file that exported with warnings

Importing H5Ps and WordPress posts must be manually tested.

legendword commented:
Wordpress Post export works well. After importing, this temporary category is added to the newly imported post (the category still exists after the Tapestry is imported):
<img width="216" alt="image" src="https://user-images.githubusercontent.com/12668055/180117258-b0cd130a-d4f7-4be6-ad2c-fca339ce5ae1.png">


The import from zip takes around 7 minutes which is fairly long for my 14.2MB export. It has two h5p nodes, one of them being a video:
<img width="753" alt="image" src="https://user-images.githubusercontent.com/12668055/180118838-73f37547-bf29-4518-8832-de89348748cb.png">
...but I still feel the import shouldn't take this long.

Below is the message after import. Everything seems to be working as expected. The Wordpress Post in the imported Tapestry correctly references the new post.
<img width="793" alt="image" src="https://user-images.githubusercontent.com/12668055/180118478-3a08e5f3-4ff2-4cf2-b340-4ef16649b64c.png">

I will review the code and try injecting errors, but since there are lots of code changes it may take a while for me to finish reviewing the code. I will post a review when I'm done.

Also could you resolve the merge conflicts when you have time @zifgu?

legendword commented:
Also I'm not sure if this is supposed to happen? There are 2 versions of each H5P after the import (see the last modified time of the first 4 H5Ps).

<img width="1267" alt="image" src="https://user-images.githubusercontent.com/12668055/180119928-99e16c15-3934-492a-af0f-0f9b65f47318.png">

zifgu commented:
> Also I'm not sure if this is supposed to happen? There are 2 versions of each H5P after the import (see the last modified time of the first 4 H5Ps).
> 
> <img alt="image" width="1267" src="https://user-images.githubusercontent.com/12668055/180119928-99e16c15-3934-492a-af0f-0f9b65f47318.png">

Hmmm... the expected behaviour is that if you originally had 2 H5Ps and exported a Tapestry with 2 H5Ps, you should have 4 H5Ps total after importing it. All media in the zip file would be created on the target site regardless of whether it originally existed there.

However, if you exported a Tapestry with 2 H5Ps and 4 H5Ps were added, that is a bug.

zifgu commented:
> Also could you resolve the merge conflicts when you have time @zifgu?

Hopefully done!


> The import from zip takes around 7 minutes which is fairly long for my 14.2MB export. It has two h5p nodes, one of them being a video:
> <img alt="image" width="753" src="https://user-images.githubusercontent.com/12668055/180118838-73f37547-bf29-4518-8832-de89348748cb.png">
> ...but I still feel the import shouldn't take this long.

In a correct scenario, H5Ps are expected to take slightly longer to import because the H5P plugin makes an HTTP request with every H5P content it adds. Also, the H5P cache is rebuilt after the import. However, I would expect both of these operations to take seconds at most.

This seems extremely long for only 2 H5Ps. Would you be able to send me the zip file you used (if GitHub supports attachments, and if not, on Slack)? I will try to reproduce it locally.

zifgu commented:
> Also I'm not sure if this is supposed to happen? There are 2 versions of each H5P after the import (see the last modified time of the first 4 H5Ps).
> 
> <img alt="image" width="1267" src="https://user-images.githubusercontent.com/12668055/180119928-99e16c15-3934-492a-af0f-0f9b65f47318.png">

You're right, this was a bug where the same H5P content is added more than once if it is used in multiple nodes. I have now fixed this so that each file in the zip is added at most once.


> The import from zip takes around 7 minutes which is fairly long for my 14.2MB export. It has two h5p nodes, one of them being a video:
> <img alt="image" width="753" src="https://user-images.githubusercontent.com/12668055/180118838-73f37547-bf29-4518-8832-de89348748cb.png">
> ...but I still feel the import shouldn't take this long.

Tested this on my computer locally and it took about 40 seconds to fetch each video, about 1 minute 30 seconds in total for the import, so I am guessing large videos are the bottleneck. If the user's local H5P plugin is messed up and they have many H5Ps whose caches need to be rebuilt, the step where we rebuild the H5P caches could also take a significant amount of time. Unfortunately, I don't know if we can do anything about this.


> Wordpress Post export works well. After importing, this temporary category is added to the newly imported post (the category still exists after the Tapestry is imported):
> <img alt="image" width="216" src="https://user-images.githubusercontent.com/12668055/180117258-b0cd130a-d4f7-4be6-ad2c-fca339ce5ae1.png">

The cause was that the ZIP file and XML file did not use the categories. I've just added a unique string to the name of each export file to help users identify which XML file and ZIP/JSON file are paired together.

I think that is everything for now, until you review the code - thanks for catching these!

legendword reviewed (commented):
Most of the code changes make sense; see some comments below. The import and export functionality seem to be still working as intended.

zifgu commented:
This commit should fix a bug where only the root node shows up for logged-out users (for example, those in Incognito mode). In `_filterImportedPerms`, the permissions object was returned as an associative array, while permission checks were expecting an object.

zifgu commented on templates/vue/src/components/common/ImportExportWarnings.vue:
```diff
@@ -0,0 +1,76 @@
+    // action is either "import" or "export"
+    action: {
+      type: String,
+      required: true,
```
Thanks! That does the job perfectly.

zifgu commented on templates/vue/src/components/TapestryMain/RootNodeButton/ImportChangelog.vue:
```diff
@@ -12,20 +12,28 @@
         </div>
+        <import-warnings :warnings="warnings" action="import"></import-warnings>
+        <p>
+          Pressing "Confirm" will reload your page.
```
That does sound better. Fixed.

zifgu commented on templates/vue/cypress/plugins/index.js:
```diff
@@ -14,5 +14,11 @@
 module.exports = (on, config) => {
   require("@cypress/code-coverage/task")(on, config)
   on("file:preprocessor", require("@cypress/code-coverage/use-browserify-istanbul"))
+  on("task", {
```
This code registers the ability to log debug messages during a Cypress test. For example, the command `cy.task("log", "Debug message")` would print `Debug message` to the console/terminal.

I added it based off the example here (https://docs.cypress.io/api/commands/task#Usage) while debugging the hanging tests on this PR. It's not strictly related to the import/export functionality, so we can definitely delete it here, but it might be a useful addition for debugging Cypress tests in the future. What do you think?

legendword replied:
Sounds good. We should put the example of how to use it somewhere on Notion so future devs know how to use it.

zifgu replied:
Good idea - I couldn't find a dedicated page for this information on Notion, so I added it as a comment above this code. Hopefully this will help.

zifgu commented on templates/vue/src/components/TapestryMain/RootNodeButton/index.vue:
```diff
@@ -88,88 +93,90 @@ export default {
-          .then(() => {
-            this.isImporting = false
+          .then(response => {
+            this.changes.permissions = new Set(response.changes.permissions)
```
Yes, the backend can return duplicate permission changes. This code is run for every permissions object we filter, so it may add the same role twice if that role was stripped from multiple permissions objects.

https://github.com/tapestry-tool/tapestry-wp/blob/c54bcf759b8f4c727c7612fb1077f3924800271e/utilities/class.tapestry-import-export.php#L87-L92

I think this property was originally a Set for the same reason. However, if it would be better, I can change it to remove duplicates in the backend and use an array here.

legendword replied:
I don't think there's a need to do it in the backend, so let's just keep it as-is for now.

zifgu commented on templates/vue/src/components/modals/SettingsModal/index.vue:
```diff
@@ -88,48 +88,81 @@
+              export, export all WordPress posts as well as the tapestry. To import,
+              first import the WordPress posts by going to Tools -> Import in the
+              WordPress dashboard, then import the tapestry.
+            </template>
```
Sounds good, I've implemented the layout using a similar approach. This is a really nice technique to extend descriptions to multiple paragraphs.

It now looks like this:
![image](https://user-images.githubusercontent.com/44929104/181622247-f0b4acb0-18f2-4002-b33e-aa2a0277959e.png)

Please let me know what you think!

legendword replied:
Looks great, thanks!

zifgu commented on utilities/class.tapestry-import-export.php:
```diff
@@ -0,0 +1,829 @@
+
+        // Look for a post whose old ID matches the current post ID
+        $query_args = array(
+            'meta_key' => 'tapestry_export_old_post_id',
```
I think you're right - thanks for catching this. I'll change it to compare against the site URL in the exported Tapestry.

zifgu commented on templates/vue/cypress/integration/import-export.spec.js:
```diff
@@ -1,28 +1,187 @@
+  beforeEach(() => {
     cy.setup()
+
     cy.intercept("PUT", "**/tapestries/**").as("import")
```
I think importing a Tapestry makes sense as `PUT /tapestries/:tapestryPostId`, since it replace an existing Tapestry with the contents of the imported Tapestry.

Using the `POST` method was the best way I could find to receive uploaded files through PHP. This is why the `import_zip` endpoint uses `POST`, but ideally, I think it should have been PUT as well.

zifgu commented:
The latest commit should fix an issue where the wrong site URL is being used to compare against the original site URL of imported WordPress posts. However, it changes the `site-url` field in the exported JSON.

The `site-url` field was originally computed from: 
https://github.com/tapestry-tool/tapestry-wp/blob/137ec94f37881ec6e93a3d6783425d30c69a1ed4/classes/class.tapestry.php#L449
https://github.com/tapestry-tool/tapestry-wp/blob/137ec94f37881ec6e93a3d6783425d30c69a1ed4/classes/class.tapestry.php#L460

It is now computed from:
https://github.com/tapestry-tool/tapestry-wp/blob/b714982432012f7c1969d1944c6d54f9fbff20be/classes/class.tapestry.php#L471

The site URL of a WordPress site is the value displayed here under Settings > General (in a single-site setup).
![image](https://user-images.githubusercontent.com/44929104/181636137-a6c2ad3d-9fef-44ad-a543-8dce9ceb2005.png)

I think this is a more accurate way of getting the site URL, rather than forming it from the permalink, because the site URL may also contain a port (e.g. `http://localhost:8888`) or elements of the path (e.g. the default site URL with WAMP looks like `http://localhost/mysitename`). I would expect these differences to mostly affect developers.

However, I can't be sure if this change will break anything else -- there may be another reason behind the original way of computing the site URL that I'm not aware of.

legendword reviewed (approved):
The new code changes look good to me.

legendword commented:
Tested importing a Tapestry from another site with exported Wordpress posts. First imported the Wordpress post which contained a temporary category. Then performed the Tapestry import from ZIP. No warnings were generated except that permissions for nonexistent users were removed. The Wordpress posts were successfully linked to the imported Wordpress post, and the Wordpress post no longer had the temporary category. Images contained in nodes were displayed correctly and stored in the uploads folder of my local site.

Bitajkr commented:
This new import/export functionality seems to be working well.

wynnset commented:
That's great to hear @Bitajkr ! We can merge this feature soon once I review the code a bit more.

Bitajkr commented:
- [ ] There's a bug with the import/export feature where the content from the tapestry does not move over to another site. This happened with the following walkthrough tapestry, which I tried to export onto another site. The content was not transferred over; however, everything else was brought over. 

Original: https://sites.tapestry-tool.com/walkthrough/tapestry/tapestry-walkthrough/#/nodes/55
Export: https://ubc.tapestry-tool.com/artsisitworkshop2022/tapestry/walkthrough-tapestry/#/nodes/10?depth=0

wynnset commented:
Hi @Bitajkr the links you have provided are on production where this new import/export feature has NOT yet been deployed. The only place where the import/export feature is currently available is on the test site. Can you please confirm whether you also have similar issues on the test site (specifically, the import-export site)?

wynnset commented:
I found a couple of issues in my testing:
1. When I export as JSON, I am getting the wordpress posts downloaded as well. The JSON format is not supposed to export any media
2. When I export as ZIP, my media is not getting downloaded. I looked at the JSON file that was exported as part of my ZIP and I think it contains these escaped backslashes that might be the culprit (these don't exist on the JSON export):
<img width="919" alt="image" src="https://user-images.githubusercontent.com/8981258/191603217-45525e09-2c61-421a-8ca0-ad54849821e3.png">

wynnset commented:
@legendword once you are done with your current priorities, please take a look at this. If I get a chance, I might try to resolve it myself. Thanks!

legendword commented:
Tested exporting a tapestry with 2 H5P nodes, and then importing that tapestry. The H5P content is imported and no warnings are generated. The two new H5P entries can also be found in the Wordpress Dashboard > H5P Content tab.

<img width="805" alt="image" src="https://user-images.githubusercontent.com/12668055/192038946-2edf4a8e-629f-4885-8ec7-b5d71929ab54.png">

legendword commented:
![image](https://user-images.githubusercontent.com/12668055/196251444-87a14f21-6a03-4369-ace7-b7ee9c50af24.png)

"Include media" button is not aligned with the "Export Tapestry" text.
